### PR TITLE
Feature/tmux apply improvements

### PR DESCRIPTION
**Pull request by yorick1989**
I was using [create-session.go](https://github.com/jubnzv/go-tmux/blob/master/examples/create-session/create-session.go) for testing this Go package but I found out that the `server.Sessions` slice and `session` struct were missing session, window and pane id/name information after initiating / applying the configuration.

I made the following changes to solve this problem:
- I updated the Apply method in the [configuration.go](https://github.com/jubnzv/go-tmux/commit/aa164d717860f9f57ce2c97c1fce212956d00855) file:
  -  It uses the `NewSession`, `NewWindow` and `SplitPane` methods for initiating the session, windows and panes.
  - Assigns the Ids and Names properly.
- The `NewSession` and `NewWindow` methods now support additional arguments to be added to the default tmux command-line agruments.
- The new `SplitPane` method was added to [window.go](https://github.com/yorick1989/go-tmux/blob/aa164d717860f9f57ce2c97c1fce212956d00855/window.go#L78C1-L100C2) for creating new panes inside of a window.
- The `ListPanes` method in [window.go](https://github.com/yorick1989/go-tmux/blob/aa164d717860f9f57ce2c97c1fce212956d00855/window.go#L45C1-L47C2) now requires the `SessionName` for pulling the panes. This because it didn't work properly if there were multiple sessions created.

I've also adjusted the following in [pane.go](https://github.com/yorick1989/go-tmux/commit/80b85bef57eab56503384d66b93fc106347eba1c):
- Added the Pipe method.
- Renamed (the struct Pipe) ID to Id for consistency.
- Added the pane Index to the Pipe struct.
- Use SessionName, WindowName and Pane Index to target pane commands.

During a newly created session, window or pane the newly created struct will also be added to the pointer-receiver.

Hopefully you find these changes valid too and will this PR be merged into your repository.

Let me know if my code need any improvements.

Thanks!

Yorick

### CODE REVIEW

**Comment by yorick1989** [comment]
I've also adjusted the following in [pane.go](https://github.com/yorick1989/go-tmux/blob/feature/tmux_apply_improvements/pane.go):
- Added the Pipe method.
- Renamed (the struct Pipe) ID to Id for consistency.
- Added the pane Index to the Pipe struct.
- Use SessionName, WindowName and Pane Index to target pane commands.

File: window.go [hunk 1]
```diff
@@ -70,3 +73,31 @@ func (w *Window) Select() error {
 	}
 	return nil
 }
+
+// Creates a pane inside this window.
+func (w *Window) SplitPane() (pane Pane, err error) {
+
```
Comment (jubnzv) [inline window.go]:
```suggestion
```

File: window.go [hunk 2]
```diff
@@ -70,3 +73,31 @@ func (w *Window) Select() error {
...
+	args := []string{
+		"split-window",
+		"-t", fmt.Sprintf("%s:%s", w.SessionName, w.Name),
+		"-c", w.StartDirectory,
+		"-F", "#{pane_id}"}
+
```
Comment (jubnzv) [inline window.go]:
```suggestion
```

File: pane.go [hunk 3]
```diff
@@ -88,19 +90,25 @@ func ListPanes(args []string) ([]Pane, error) {
...
+		}
+
+		paneIndex, errAtoi := strconv.Atoi(result[8])
 		if errAtoi != nil {
 			return nil, errAtoi
 		}
 
 		panes = append(panes, Pane{
-			ID:          paneIndex,
+			Id:          paneId,
```
Comment (jubnzv) [inline pane.go]: We should not rename public entries this to don't break backward compatibility for library's users.

```suggestion
			ID:          paneId,
```

File: configuration.go [hunk 4]
```diff
@@ -50,94 +49,96 @@ func (c *Configuration) Apply() error {
...
-			_, _, err_exec := RunCmd(args)
-			if err_exec != nil {
-				return err_exec
+				if len(w.StartDirectory) != 0 {
+					extra_args = append(extra_args, "-c", w.StartDirectory)
+				}
+
+				// Create a new window
+				window, err := s.NewWindow(w.Name, extra_args...)
+
```
Comment (jubnzv) [inline configuration.go]:
```suggestion
```

File: configuration.go [hunk 5]
```diff
@@ -50,94 +49,96 @@ func (c *Configuration) Apply() error {
...
 
+			w.SessionName = s.Name
+			w.SessionId = s.Id
+
 			// Setup panes for created window
-			for idx := range w.Panes {
+			orig_panes := w.Panes
+
+			w.Panes, _ = w.ListPanes()
+
```
Comment (jubnzv) [inline configuration.go]:
```suggestion
```

File: configuration.go [hunk 6]
```diff
@@ -50,94 +49,96 @@ func (c *Configuration) Apply() error {
...
+				w.Id = window.Id
 			}
 
+			w.SessionName = s.Name
+			w.SessionId = s.Id
+
 			// Setup panes for created window
-			for idx := range w.Panes {
+			orig_panes := w.Panes
+
```
Comment (jubnzv) [inline configuration.go]:
```suggestion
```

File: configuration.go [hunk 7]
```diff
@@ -50,94 +49,96 @@ func (c *Configuration) Apply() error {
...
-						"-t", winId,
-						"-c", windowStartDirectory,
-					}
-					_, _, err_exec := RunCmd(args)
-					if err_exec != nil {
-						return err_exec
+				if idx > 0 {
+					// Create a new pane
+					pane, err := w.SplitPane()
+
```
Comment (jubnzv) [inline configuration.go]:
```suggestion
```

File: configuration.go [hunk 8]
```diff
@@ -50,94 +49,96 @@ func (c *Configuration) Apply() error {
...
 					return err_exec
 				}
 			}
+
+			win = append(win, w)
 		}
+
+		c.Sessions[si].Windows = win
+		c.Sessions[si] = s
+
```
Comment (jubnzv) [inline configuration.go]:
```suggestion
```

File: pane.go [hunk 9]
```diff
@@ -13,21 +13,22 @@ import (
 // Represent a tmux pane:
 // https://github.com/tmux/tmux/wiki/Getting-Started#sessions-windows-and-panes
 type Pane struct {
-	ID          int
+	Id          int
```
Comment (jubnzv) [inline pane.go]:
```suggestion
	ID          int
```

File: pane.go [hunk 10]
```diff
@@ -13,21 +13,22 @@ import (
...
+	Index       int
 }
 
 // Creates a new pane object.
 func NewPane(id int, sessionId int, sessionName string, windowId int,
 	windowName string, windowIndex int, active bool,
 ) *Pane {
 	return &Pane{
-		ID:          id,
+		Id:          id,
```
Comment (jubnzv) [inline pane.go]:
```suggestion
		ID:          id,
```

**Review by jubnzv — CHANGES_REQUESTED** [review]
Hey @yorick1989,

Thank you for the contribution.

I see the point; it should work as expected. I have only a few styling comments and suggest not breaking backward compatibility by renaming things.

Otherwise, LGTM.

**Comment by yorick1989** [comment]
> Hey @yorick1989,
> 
> Thank you for the contribution.
> 
> I see the point; it should work as expected. I have only a few styling comments and suggest not breaking backward compatibility by renaming things.
> 
> Otherwise, LGTM.

Sounds good. Let me know if it can be merged with your repository this way =).

Thanks!

Yorick

**Review by yorick1989 — COMMENTED** [review]
All looks good! Hopefully this is good enough :).

**Comment by jubnzv** [comment]
Thank you!

Could we also make `golang-ci` happy? https://github.com/jubnzv/go-tmux/actions/runs/10282424723/job/28466439660?pr=19#step:4:25

**Comment by yorick1989** [comment]
> Thank you!
> 
> Could we also make `golang-ci` happy? https://github.com/jubnzv/go-tmux/actions/runs/10282424723/job/28466439660?pr=19#step:4:25

Good finding. I fixed the int to string converter @ https://github.com/jubnzv/go-tmux/pull/19/commits/30bebe31e254369249c519a3b18c083c862fe31c . I'd also run the linter myself:
```bash
[yorick@yorick-pc go-tmux]$ golangci-lint run -c .golangci.yml
[yorick@yorick-pc go-tmux]$ 
```
No errors at all.

Please verify and hopefully this will be enough to pass trough.

Thanks!

**Pull request merged**